// vueify/lib/compilers/babel.js
var fs = require('fs')
var path = require('path')
var json = require('json5')
var assign = require('object-assign')
var ensureRequire = require('../ensure-require')

var defaultBabelOptions = {
  presets: ['@babel/preset-env'],
  plugins: [
    '@babel/plugin-transform-runtime',
    // [
    //   '@babel/plugin-transform-runtime',
    //   {
    //     corejs: 3,
    //   }
    // ]
  ],
};

if(
  process.env.NODE_ENV === 'staging' ||
  process.env.NODE_ENV === 'production'
  || (
    process.env.NODE_ENV === 'development' &&
    process.env.MODE !== 'fast'
  )
) {
  defaultBabelOptions.plugins = defaultBabelOptions.plugins.concat([
    ["module-resolver", {
      "root": ["."],
      "alias": {
      "~jsPath": "./src/js",
      "~jsBasePath": "./src/js/base",
      "~jsLibPath": "./src/js/lib",
      "~jsDefinePath": "./src/js/define",
      "~jsHelperPath": "./src/js/helper",
      "~jsPartialPath": "./src/js/partial",
      "~jsLandingPath": "./src/js/landing"
      }
    }],
    "dynamic-import-split-require",
  ]);
}

var babelRcPath = path.resolve(process.cwd(), '.babelrc')
var babelOptions = fs.existsSync(babelRcPath)
  ? getBabelRc() || defaultBabelOptions
  : defaultBabelOptions

function getBabelRc () {
  var rc
  try {
    rc = json.parse(fs.readFileSync(babelRcPath, 'utf-8'))
  } catch (e) {
    throw new Error('[vueify] Your .babelrc seems to be incorrectly formatted.')
  }
  return rc
}

module.exports = function (raw, cb, compiler, filePath) {
  if ((compiler.options.babel || babelOptions) === defaultBabelOptions) {
    try {
      ensureRequire('babel', ['@babel/preset-env', '@babel/plugin-transform-runtime'])
    } catch (e) {
      console.error(e.message)
      console.error(
        '\n^^^ You are seeing this because you are using Vueify\'s default babel ' +
        'configuration. You can override this with .babelrc or the babel option ' +
        'in vue.config.js.'
      )
    }
  }

  try {
    var babel = require('@babel/core')
    var options = assign({
      comments: false,
      filename: filePath,
      sourceMaps: compiler.options.sourceMap
    }, compiler.options.babel || babelOptions)
    var res = babel.transform(raw, options)
  } catch (err) {
    return cb(err)
  }
  cb(null, res)
}
